cli/js/permissions.ts

```diff
@@ -2,38 +2,59 @@
 import * as dispatch from "./dispatch.ts";
 import { sendSync } from "./dispatch_json.ts";
 
-/** Permissions as granted by the caller */
-export interface Permissions {
-  read: boolean;
-  write: boolean;
-  net: boolean;
-  env: boolean;
-  run: boolean;
-  hrtime: boolean;
-  // NOTE: Keep in sync with src/permissions.rs
+/** Permissions as granted by the caller
+ * See: https://w3c.github.io/permissions/#permission-registry
+ */
+export type PermissionName =
+  | "read"
+  | "write"
+  | "net"
+  | "env"
+  | "run"
+  | "hrtime";
+// NOTE: Keep in sync with cli/permissions.rs
+
+/** https://w3c.github.io/permissions/#status-of-a-permission */
+export type PermissionState = "granted" | "denied" | "prompt";
+
+/** See: https://w3c.github.io/permissions/#permission-descriptor */
+interface SimplePermissionDescriptor {
+  name: "env" | "run" | "hrtime";
 }
 
-export type Permission = keyof Permissions;
+interface NetPermissionDescriptor {
+  name: "net";
+  url?: string;
+}
 
-/** Inspect granted permissions for the current program.
- *
- *       if (Deno.permissions().read) {
- *         const file = await Deno.readFile("example.test");
- *         // ...
- *       }
- */
-export function permissions(): Permissions {
-  return sendSync(dispatch.OP_PERMISSIONS) as Permissions;
+interface PathPermissionDescriptor {
+  name: "read" | "write";
+  path?: string;
 }
 
-/** Revoke a permission. When the permission was already revoked nothing changes
- *
- *       if (Deno.permissions().read) {
- *         const file = await Deno.readFile("example.test");
- *         Deno.revokePermission('read');
- *       }
- *       Deno.readFile("example.test"); // -> error or permission prompt
- */
-export function revokePermission(permission: Permission): void {
-  sendSync(dispatch.OP_REVOKE_PERMISSION, { permission });
+type PermissionDescriptor =
+  | SimplePermissionDescriptor
+  | PathPermissionDescriptor
+  | NetPermissionDescriptor;
```

Comment:
👍

Opinion: I think it would be better to give each permission its own interface so they are arbitrarily extensible e.g. #3137.

Reply:
Makes sense 👍  I separated descriptor types for each permission.